guide/secure_secret_management.md

```diff
@@ -0,0 +1,110 @@
+
+## Using Environment Variables
+First, and least secure, is the ability to set environment variables at deploy time. For example, you might want to set the `NODE_ENV` or `DEBUG` variable.  Setting the `NODE_ENV` in the stack file `samples.yml`
+```yaml
+provider:
+  name: faas
+  gateway: http://localhost:8080
+
+functions:
+  nodehelloenv:
+    lang: Dockerfile
+    skip_build: true
+    image: functions/nodehelloenv:latest
+    environment:
+        NODE_ENV: production
+```
+
+You can then deploy and invoke the function via the `faas-cli` using
+```sh
+$ faas-cli deploy -f ./samples.yml
+$ fass-cli invoke nodehelloenv
+Hello from a production machine
+```
+
+Notice that it is using the value of `NODE_ENV` from the stack file, the default is is `dev`.
+
+
+## Using Swarm Secrets
+For sensitive value we can leverage the [Docker Swarm Secrets](https://docs.docker.com/engine/swarm/secrets/) feature to safely store and give our functions access to the needed values. Using secrets is a two step process.  Take the [ApiKeyProtected](../sample-functions/ApiKeyProtected) example function, when we deploy this function we provide a secret key that it uses to authenticate requests to it.  First we must add a secret to the swarm
+
+```sh
+docker secret create secret_api_key ~/secrets/secret_api_key.txt
+```
+
+where `~/secrets/secret_api_key.txt` is a simple text file that might look like this
+
+```txt
+R^YqzKzSJw51K9zPpQ3R3N
+```
+
+Equivalently, you can pipe the value to docker via stdin like this
+
+```sh
+echo "R^YqzKzSJw51K9zPpQ3R3N" | docker secret create secret_api_key -
+```
+
+Now, with the secret defined, we can deploy the function like this
+
+
+```sh
+$ echo "R^YqzKzSJw51K9zPpQ3R3N" | docker secret create secret_api_key -
+$ faas-cli deploy -f ./samples.yml --secret secret_api_key
+$ curl -H "Content-Type: application/json" \
+     -X POST \
+     -H "X-Api-Key: R^YqzKzSJw51K9zPpQ3R3N" \
+     -d '{}' \
+     http://localhost:8080/function/protectedapi
+
+Unlocked the function!
+
+$ curl -H "Content-Type: application/json" \
+     -X POST \
+     -H "X-Api-Key: wrong_key" \
+     -d '{}' \
+     http://localhost:8080/function/protectedapi
+
+Access denied!
+```
+
+Your `samples.yml` stack file looks like this
+```yaml
+provider:
+  name: faas
+  gateway: http://localhost:8080
+
+functions:
+  protectedapi:
+    lang: Dockerfile
+    skip_build: true
+    image: functions/api-key-protected:latest
+```
+
+Note that unlike the `envVars` in the first example, we do not provide the secret value, just a list of names: `"secrets": ["secret_api_key"]`. The secret value has already been securely stored in the Docker swarm.  One really great result of this type of configuration is that you can simplify your function code by always referencing the same secret name, no matter the environment, the only change is how the environments are configured.
+
```

Comment:
I haven't been able to get the secrets to work from the yml file... Here, you describe adding a "list of names", but there's no mention of it in the example above. 
I've tried a yml style list:
```yaml
functions:
  secret_test:
    lang: node
    handler: ./secret_test
    image: secret_test:test
    secrets: 
      - API_SECRET_KEY
```

I've also tried matching your inline example format:
```yaml
functions:
  secret_test:
    lang: node
    handler: ./secret_test
    image: secret_test:test
    secrets: ["API_SECRET_KEY"]
```
_with and without the quotes just in case_

both with no luck.

Deploying using the cli _did_ work though
`faas-cli deploy -f secret_test.yml --secret API_SECRET_KEY`

or, am I misreading this in that you don't define the secret in the yml file, but only through the cli?

Reply:
Your example yml files are definitely what is intended and I should update it. When you say deploying with the CLI worked, do you mean that `faas-cli deploy -f secret_test.yml` did not work?

Reply:
Right. 
So, setting the secrets _only_ using the yml file and deploying with the cli `faas-cli deploy -f secret_test.yml` I was not able to read the secret. This is the same with deploying via the UI (no secret)

Deploying with the cli adding the additional secrets flag `faas-cli deploy -f secret_test.yml --secret API_SECRET_KEY` did work and the secret is available

Reply:
Seems like it might be an error in the `fass-cli` parsing the stack file, I will take another look

Reply:
@BurtonR , are you updated to the latest version of the faas-cli? The secrets parsing was added fairly recently

Reply:
Looks like release `0.5.1` should support both the stack file and the cli args

Reply:
For the record, I just tried this out, deployed the `apikeyprotected` function with the secret specified in the stack yaml, and the secret is showing up in the container (at `/run/secrets/<secretname>`). @BurtonR , are you seeing an issue with the secret not showing up in the container at all, or are you seeing the secret show up in the container but not be able to read the secret from its location in `/run/secrets/`?

Reply:
@ericstoekl not sure what the difference is between being in the container, and being in the secrets location...

When testing this, I created a simple node function to read the secret from the `/run/secrets/` location and it was throwing an error that the file didn't exist (the secret wasn't available in the function)

Also, the cli 0.5.1 hasn't been released yet. It's still marked as Pre-Release, which is why I don't have it yet. I'll grab the binary and try this out again. 

The cli version compatibility should be a note in the guide 😉

Reply:
Alright, so, using the faas-cli version 0.5.1 with the secrets defined in the yml file:
```yaml
functions:
  secret_test:
    lang: node
    handler: ./secret_test
    image: secret_test:test
    secrets: 
      - API_SECRET_KEY
```

The secret _is_ available and working as I expect it to.

Once the guide is updated to include these couple of little things I'd call this good to go! 
Nice work @LucasRoesler!

Reply:
I have updated the guide with a note at the beginning of the section using Swarm secrets.

Reply:
Thanks for testing this out @BurtonR